slither/tools/read_storage/read_storage.py

```diff
@@ -0,0 +1,521 @@
+import sys
+import json
+import logging
+from math import floor
+
+from typing import Any, Tuple, Union, List
+
+try:
+    from web3 import Web3
+    from eth_typing.evm import ChecksumAddress
+    from eth_abi import decode_single, encode_abi
+    from eth_utils import keccak
+    from .utils import (
+        is_array,
+        is_mapping,
+        is_struct,
+        is_user_defined_type,
+        get_offset_value,
+        get_storage_data,
+        coerce_type,
+    )
+except ImportError:
+    print("ERROR: in order to use slither-read-storage, you need to install web3")
+    print("$ pip3 install web3 --user\n")
+    sys.exit(-1)
+
+from slither.core.declarations import Contract, StructureContract
+from slither.core.solidity_types.type import Type
+from slither.core.variables.state_variable import StateVariable
+from slither.core.variables.structure_variable import StructureVariable
+
+
+logging.basicConfig()
+logger = logging.getLogger("Slither-read-storage")
+logger.setLevel(logging.INFO)
+
+
+class SlitherReadStorageException(Exception):
+    pass
+
+
+# pylint: disable=too-many-arguments
+def _all_struct_variables(var, contracts, address, rpc_url, storage_address=None, key=None):
+    """Retrieves all members of a struct."""
+    if isinstance(var.type.type, StructureContract):
+        struct_elems = var.type.type.elems_ordered
+    else:
+        struct_elems = var.type.type.type.elems_ordered
+    data = {}
+    for _, elem in enumerate(struct_elems):
+        slot, val, type_string = get_storage_slot_and_val(
+            contracts,
+            address,
+            var.name,
+            rpc_url,
+            storage_address,
+            key=key,
+            struct_var=elem.name,
+        )
+        data[elem.name] = {"slot": slot, "value": val, "type_string": type_string}
+
+    return data
+
+
+# pylint: disable=too-many-branches,too-many-locals,too-many-nested-blocks
```

Comment:
I think we can split this function into sub-functions to not remove all these warnings